src/lib.rs

```diff
@@ -157,61 +162,98 @@ impl Keysym {
-            | key::KP_Tab | key::KP_Enter | key::KP_Multiply..=key::KP_9
-            | key::KP_Equal | key::Delete
-        ) {
+        // Also check for directly encoded unicode codepoints.
+
+        // Exclude surrogates: they are invalid in UTF-32.
+        // See https://www.unicode.org/versions/Unicode15.0.0/ch03.pdf#G28875
+        // for further details.
+        if matches!(keysym, 0x0100d800..=0x0100dfff) {
             return None;
         }
 
-        // Convert to ASCII by converting the low byte.
-        let mut ascii_key = match (keysym, high_bytes) {
-            (key::KP_Space, _) => b' ',
-            (_, 0xFF) => (keysym & 0x7F) as u8,
-            _ => keysym as u8,
-        };
-
-        // Apply the control key if it makes sense.
-        if has_control_key {
-            match ascii_key {
-                b'@'..=126 | b' ' => {
-                    ascii_key &= 0x1F;
-                }
-                b'2' => {
-                    ascii_key = b'\0';
-                }
-                b'3'..=b'7' => {
-                    ascii_key -= b'3' - 27;
-                }
-                b'8' => {
-                    ascii_key = 127;
-                }
-                b'/' => {
-                    ascii_key = b'_' & 0x1F;
-                }
-                _ => {}
-            }
+        // In theory, this is supposed to start from 0x100100, such that the ASCII
+        // range, which is already covered by 0x00-0xff, can't be encoded in two
+        // ways. However, changing this after a couple of decades probably won't
+        // go well, so it stays as it is.
+        if matches!(keysym, 0x01000000..=0x0110ffff) {
+            return char::from_u32(keysym - 0x01000000);
+        }
+
+        // Search the main table.
+        KEYSYMTAB
+            .binary_search_by(|pair| pair.keysym.cmp(&(keysym as u16)))
+            .ok()
+            .and_then(|idx| char::from_u32(KEYSYMTAB[idx].ucs as u32))
+    }
+
+    /// Create a [`Keysym`] from the given `char`.
+    ///
+    /// This function replecates the `xkb_utf32_to_keysym` behavior.
+    pub fn from_char(ch: char) -> Self {
+        let ucs = ch as u32;
+
+        // First check for Latin-1 characters (1:1 mapping).
+        if matches!(ucs, 0x0020..=0x007e) || matches!(ucs, 0x00a0..=0x00ff) {
+            return Self::new(ucs);
+        }
+
+        // Special keysyms.
+        if (key::BackSpace & 0x7f..=key::Clear & 0x7f).contains(&ucs)
+            || ucs == key::Return & 0x7f
+            || ucs == key::Escape & 0x7f
+        {
+            return Self::new(ucs | 0xff00);
+        }
+
+        if ucs == key::Delete & 0x7f {
+            return Self::new(key::Delete);
+        }
+
+        // Unicode non-symbols and code points outside Unicode planes.
+        if matches!(ucs, 0xd800..=0xdfff)
+            || matches!(ucs, 0xfdd0..=0xfdef)
+            || ucs > 0x10ffff
+            || (ucs & 0xfffe == 0xfffe)
+        {
+            return NO_SYMBOL;
```

Comment:
Maybe this function should return `Option<Keysym>` instead? Feels more idiomatic than checking for `NO_SYMBOL`.

(Although, this makes me think, should we define the types in this crate using the `NonZero*` family?)

Reply:
Well, I've followed it exactly, and given that we have `NO_SYMBOL` and that we have a constructors not checking for it, I've decided to use `NO_SYMBOL`. I've originally used `Option`, but once I've seen that you don't check for `0` in `new` and that you have `NO_SYMBOL` I did what you see.